python/feathub/metric_stores/prometheus_metric_store.py

```diff
@@ -0,0 +1,220 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import collections
+from typing import Any, Dict, List, Optional, OrderedDict, Tuple
+
+from feathub.common.config import ConfigDef
+from feathub.common.utils import generate_random_name
+from feathub.feature_tables.sinks.prometheus_sink import PrometheusSink
+from feathub.feature_tables.sinks.sink import Sink
+from feathub.feature_views.derived_feature_view import DerivedFeatureView
+from feathub.feature_views.feature import Feature
+from feathub.feature_views.sliding_feature_view import SlidingFeatureView
+from feathub.metric_stores.metric import Metric
+from feathub.metric_stores.metric_store import MetricStore
+from feathub.metric_stores.metric_store_config import (
+    METRIC_STORE_PREFIX,
+)
+from feathub.metric_stores.metric_store_config import (
+    MetricStoreConfig,
+)
+from feathub.processors.materialization_descriptor import MaterializationDescriptor
+from feathub.table.table_descriptor import TableDescriptor
+
+PROMETHEUS_METRIC_STORE_PREFIX = METRIC_STORE_PREFIX + "prometheus."
+
+SERVER_URL_CONFIG = PROMETHEUS_METRIC_STORE_PREFIX + "server_url"
+SERVER_URL_DOC = "The PushGateway server URL including scheme, host name, and port."
+
+DELETE_ON_SHUTDOWN_CONFIG = PROMETHEUS_METRIC_STORE_PREFIX + "delete_on_shutdown"
+DELETE_ON_SHUTDOWN_DOC = """
+Whether to delete metrics from Prometheus when the job finishes. When set to true,
+Feathub will try its best to delete the metrics but this is not guaranteed.
+"""
+
+
+prometheus_metric_store_config_defs: List[ConfigDef] = [
+    ConfigDef(
+        name=SERVER_URL_CONFIG,
+        value_type=str,
+        description=SERVER_URL_DOC,
+    ),
+    ConfigDef(
+        name=DELETE_ON_SHUTDOWN_CONFIG,
+        value_type=bool,
+        description=DELETE_ON_SHUTDOWN_DOC,
+        default_value=True,
+    ),
+]
+
+
+class PrometheusMetricStoreConfig(MetricStoreConfig):
+    def __init__(self, props: Dict[str, Any]) -> None:
+        super().__init__(props)
+        self.update_config_values(prometheus_metric_store_config_defs)
+
+
+# TODO: Allow users to configure feature name pattern.
+class PrometheusMetricStore(MetricStore):
+    """
+    A metric store reporting metrics to Prometheus through push gateway. For
+    user-defined Metrics, they would be reported in the following format:
+
+    - metric name: "{namespace}_{feature_name}_{metric_type}".
+    - tags: the metric tags along with the following tags.
```

Comment:
Since Prometheus use `label` instead of `tag`, should we use label as appropriate for Prometheus-specific classes?